blocks/log.py

```diff
@@ -1,274 +1,117 @@
-
-    In addition to the set of records displaying training dynamics, a
-    training log has a status object whose attributes form the state of
-
-    Another related concept is a row of the log, which is a set of record
-    sharing the same time component. The log interface has a few routines
-    to allow convenient access to the rows.
-
-    """
-    @abstractmethod
-    def get_status(self):
-        """Returns the training status.
+        return iter(self._log)
 
-        Returns
-        -------
-            An instance of :class:`AbstractTrainingStatus`, whose
-            attributes contain the information regarding the status of
-            the training process.
-
-        """
-        pass
+    def __len__(self):
+        return len(self._log)
 
     @property
-    def status(self):
-        """A convenient way to access the status."""
-        return self.get_status()
-
-    @abstractmethod
-    def set_default_value(self, key, value):
-        """Sets a default value for 'key'."""
-        pass
-
-    @abstractmethod
-    def get_default_value(self, key):
-        """Returns the default value set for the 'key'.
-
-        Returns
-        -------
-        The default value for the `key`, or ``None`` if not set.
-
-        """
-        pass
-
-    def add_record(self, time, key, value):
-        """Adds a record to the log.
-
-        If `value` equals to the default value for the `key`, nothing is
-        done.
-
-        """
-        self._check_time(time)
-        default_value = self.get_default_value(key)
-        if value != default_value:
-            self._add_record(time, key, value)
-
-    @abstractmethod
-    def _add_record(self, time, key, value):
-        """Adds a record to the log.
-
-        The implementation method to be overridden.
-
-        """
-        pass
-
-    def fetch_record(self, time, key):
-        """Fetches a record from the log.
-
-        If no such 'key' for the `time` is found or if the value for the
-        key is ``None``, the default value for the 'key' is returned.
-
-        """
-        self._check_time(time)
-        value = self._fetch_record(time, key)
-        if value is not None:
-            return value
-        return self.get_default_value(key)
-
-    @abstractmethod
-    def _fetch_record(self, time, key):
-        """Fetches a record from the log.
-
-        The implementation method to be overridden.
+    def current_entry(self):
+        return self[self.status['iterations_done']]
 
-        """
-        pass
+    @property
+    def previous_entry(self):
+        return self[self.status['iterations_done'] - 1]
 
-    def __getitem__(self, time):
-        self._check_time(time)
-        return TrainingLogRow(self, time)
 
-    @property
-    def current_row(self):
-        return self[self.status.iterations_done]
+class TrainingStatus(Mapping):
+    def __init__(self, exclude=None):
+        self._status = {'iterations_done': 0, 'epochs_done': 0}
```

Comment:
`epoch_finished=True` + a method for the log to fetch a column as a dataframe would do the job:

```
epoch_ends = log.get_column('epoch_finished').dropna().index
```

But it orthogonal to two other tasks this PR tries to solve and I think can be postponed. Small PRs are a way easier to review.